fetchmail_wakeup.7
.\" manual page for fetchmail_wakeup
.\" Copyright (C) 2009 Peter Marschall <peter@adpm.de>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH FETCHMAIL_WAKEUP 7 "25 July 2009" "" "Dovecot Plugins"

.SH NAME
fetchmail_wakeup \- Dovecot plugin to trigger fetchmail wakeups

.SH DESCRIPTION
fetchmail_wakeup is a plugin for Dovecot 1.1.
.PP
It intercepts the following IMAPv4 commands to fetch mail from the server:
.TP
.B STATUS
client polling for new mail
.TP
.B IDLE
client telling the server to push new mail
.PP
and tries to wake up a running fetchmail daemon before performing
the command.
.PP
fetchmail_wakeup supports two different ways of waking up a fetchmail
daemon:
.TP 3
*
sending the SIGUSR1 signal to a PID stored in the first line of a PID file
.TP
*
executing a command
.PP
As both, the PID file and the command to execute, can be configured in
Dovecot's configuration file \fIdovecot.conf\fP, you are not restricted to
fetchmail, but can execute or send the SIGUSR1 signal to any command you want.


.SH CONFIGURATION
After the plugin library is installed and ready to be used, Dovecot's
configuration file, usually \fI/etc/dovecot/dovecot.conf\fP, needs to be adapted
to make Dovecot use the plugin:

First, make Dovecot aware of the plugin.
This happens in dovecot.conf's \fBprotocol imap\fP section and comprises
.TP 3
*
setting the mail_plugin_dir variable (if not already set)
.TP
*
extending the mail_plugins variable by the plugin's name "fetchmail_wakeup"
.PP
.nf
\fBprotocol imap\fP {
    ...

    \fBmail_plugins\fP = \fBfetchmail_wakeup\fP
    \fBmail_plugin_dir\fP = /usr/lib/dovecot/modules/imap

    ...
}
.fi
.PP
Then, set the plugin's configuration options.
This is done in dovecot.conf's \fBplugin\fP section.

fetchmail_wakeup supports three configuration options:
.TP
.BR fetchmail_interval = \fINUMBER\fP
Set minimal interval between two fetchmail invocations to \fINUMBER\fP seconds.
If it is not given, the interval defaults to \fB60\fP.
.TP
.BR fetchmail_helper = \fICOMMAND\fP
Execute \fICOMMAND\fP to either start fetchmail (or any other mail fetching tool),
or to awaken a running fetchmail daemon (or any other mail fetching tool).
.TP
.BR fetchmail_pidfile = \fINAME\fP
Use \fINAME\fP as the file to read the process ID (PID) of a running fetchmail
instance from, and awaken this instance by sending it the \fBSIGUSR1\fP signal.
.PP
If fetchmail_helper is given, it takes precedence and fetchmail_pidfile
is ignored.
.PP
.nf
\fBplugin\fP {
    ...

    ## fetchmail_wakeup plugin: awaken fetchmail on STATUS & IDLE commands

    # minimal interval (in seconds) between two awakenings of fetchmail
    # (this setting is optional and defaults to 60)
    \fBfetchmail_interval\fP = 60

    # a helper program to notify fetchmail accordingly
    \fBfetchmail_helper\fP = /usr/bin/awaken-fetchmail

    # pid file of a running fetchmail instance
    \fBfetchmail_pidfile\fP = %h/fetchmail.pid

    ...
}
.fi
.PP
As the example above shows, it is even possible to use Dovecot's variables
in the configuration options, and thus awaken the private fetchmail daemons
of multiple users with this configuration.
.PP
Now restart your dovecot daemon and enjoy the comfort of being able to control
fetchmail by simply fetching mail from the server in your IMAP client.


.SH ERRORS
fetchmail_wakeup reports its errors via Dovecot's regular logging mechanisms.

.SH FILES
.I dovecot.conf
.br
.I lib_fetchmail_wakeup_plugin.so


.SH NOTES
Although some kind of runtime-configurable rate-limiting is supported
using \fIfetchmail_interval\fP,
this module is not considered suitable for large scale installations.
But those installations do not rely on fetchmail for pulling in
new mail from mail providers anyway.


.SH SEE ALSO
.BR dovecot (1),
.BR fetchmail (1)




fetchmail_wakeup.c
/*
 * Fetchmail notification IMAP plugin for Dovecot
 *
 * Copyright (C) 2007 Guillaume Chazarain <guichaz@yahoo.fr>
 * - original version named wake_up_fetchmail.c
 *
 * Copyright (C) 2009 Peter Marschall <peter@adpm.de>
 * - adaptions to dovecot 1.1
 * - rename to fetchmail_wakeup.c
 * - configuration via dovecot.config
 *
 * License: LGPL v2.1
 *
 */

#include <sys/types.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <limits.h>
#include "lib/lib.h"
#include "lib/imem.h"
#include "imap/commands.h"


#define FETCHMAIL_PIDFILE	"/var/run/fetchmail/fetchmail.pid"
#define FETCHMAIL_INTERVAL	60

/*
 * delay (in seconds) between two invocations of fetchmail
 */
static int fetchmail_interval = FETCHMAIL_INTERVAL;


/*
 * Dovecot's real "IDLE" function
 */
static struct command *orig_cmd_idle_ptr;
static struct command orig_cmd_idle;
/*
 * Dovecot's real "STATUS" function
 */
static struct command *orig_cmd_status_ptr;
static struct command orig_cmd_status;


/*
 * Don't bother waking up fetchmail too often
 */
static bool ratelimit(long interval)
{
	static struct timeval last_one;
	struct timeval now;
	long millisec_delta;

	if (gettimeofday(&now, NULL))
		return FALSE;

	millisec_delta = ((now.tv_sec - last_one.tv_sec) * 1000000 +
	                  now.tv_usec - last_one.tv_usec) / 1000;
	if (millisec_delta > interval * 1000) {
		last_one = now;
		return FALSE;
	}

	return TRUE;
}


/*
 * Send a signal to fetchmail or call a helper to awaken fetchmail
 */
static void fetchmail_wakeup(struct client_command_context *cmd)
{
	/* read config variables depending on the session */
	char *fetchmail_helper = getenv("FETCHMAIL_HELPER");
	char *fetchmail_pidfile = getenv("FETCHMAIL_PIDFILE");

	if (ratelimit(fetchmail_interval))
		return;

	/* if a helper application is defined, then call it */
	if ((fetchmail_helper != NULL) && (*fetchmail_helper != '\0')) {
		pid_t pid;
		int status;
		char *const *argv;

		switch (pid = fork()) {
			case -1:	// fork failed
				i_warning("fetchmail_wakeup: fork() failed");
				return;
			case 0:		// child
				argv = (char *const *) t_strsplit_spaces(fetchmail_helper, " ");
				if ((argv != NULL) && (*argv != NULL)) {
					execv(argv[0], argv);
					i_warning("fetchmail_wakeup: execv(%s) failed: %s",
						argv[0], strerror(errno));
					exit(1);
				}
				else {
					i_warning("fetchmail_wakeup: illegal fetchmail_helper");
					exit(1);
				}
			default:	// parent
				waitpid(pid, &status, 0);
		}
	}
	/* otherwise if a pid file name is given, signal fetchmail with that pid */
	else if ((fetchmail_pidfile != NULL) && (*fetchmail_pidfile != '\0')) {
		FILE *pidfile = fopen(fetchmail_pidfile, "r");
		if (pidfile) {
			pid_t pid = 0;

			if ((fscanf(pidfile, "%d", &pid) == 1) && (pid > 1))
				kill(pid, SIGUSR1);
			else
				i_warning("fetchmail_wakeup: error reading valid pid from %s",
					fetchmail_pidfile);
			fclose(pidfile);
		}
		else {
			i_warning("fetchmail_wakeup: error opening %s",
				 fetchmail_pidfile);
		}
	}
	/* otherwise warn on missing configuration */
	else {
		i_warning("fetchmail_wakeup: neither fetchmail_pidfile nor fetchmail_helper given");
	}
}


/*
 * Our "IDLE" wrapper
 */
static bool new_cmd_idle(struct client_command_context *cmd)
{
	fetchmail_wakeup(cmd);
	return orig_cmd_idle.func(cmd);
}

/*
 * Our "STATUS" wrapper
 */
static bool new_cmd_status(struct client_command_context *cmd)
{
	fetchmail_wakeup(cmd);
	return orig_cmd_status.func(cmd);
}


/*
 * Plugin init: remember dovecot's "IDLE" and "STATUS" functions and add our own
 * in place
 */
void fetchmail_wakeup_plugin_init(void)
{
	char *str;

	/* parse global config variable "fetchmail_interval" */
	str = getenv("FETCHMAIL_INTERVAL");
	if (str != NULL) {
		if (is_numeric(str, '\0')) {
			long value = strtol(str, NULL, 10);

			if (value > 0)
				fetchmail_interval = value;
			else
				i_warning("fetchmail_interval must be a positive number");
		}
		else {
			i_warning("fetchmail_interval must be a positive number");
		}
	}

	/* replace "IDLE" command handler by our own */
	orig_cmd_idle_ptr = command_find("IDLE");
	if (orig_cmd_idle_ptr)
		memcpy(&orig_cmd_idle, orig_cmd_idle_ptr, sizeof(struct command));
	command_unregister("IDLE");
	command_register("IDLE", new_cmd_idle, orig_cmd_idle.flags);

	/* replace "STATUS" command handler by our own */
	orig_cmd_status_ptr = command_find("STATUS");
	if (orig_cmd_status_ptr)
		memcpy(&orig_cmd_status, orig_cmd_status_ptr, sizeof(struct command));
	command_unregister("STATUS");
	command_register("STATUS", new_cmd_status, orig_cmd_status.flags);
}

/*
 * Plugin deinit: restore dovecot's "IDLE" and "STATUS" functions
 * The command name is dupped, for its memory location to be accessible even
 * when the plugin is unloaded
 */
void fetchmail_wakeup_plugin_deinit(void)
{
	if (orig_cmd_idle_ptr) {
		command_unregister("IDLE");
		command_register(orig_cmd_idle.name, orig_cmd_idle.func, orig_cmd_idle.flags);
	}

	if (orig_cmd_status_ptr) {
		command_unregister("STATUS");
		command_register(orig_cmd_status.name, orig_cmd_status.func, orig_cmd_status.flags);
	}
}

/* EOF */
